File: decorator/java/ComporessionDecorator.java
package Decorator;

// Yeni davranışların tanımlandığı sınıftır.
// UML diyagramındaki BaseDecorator (BaseDataSourceDecorator) sınıfından türer.
// UML diyagramındaki ConcreteDecorator sınıfına denk gelmektedir.

class ComporessionDecorator extends BaseDataSourceDecorator {
    public ComporessionDecorator(IDataSource dataSource) {
        super(dataSource);
    }

    @Override
    public String getFileName() {
        return super.dataSource.getFileName();
    }

    @Override
    public void readData() {
        // Veriyi okurken referansını tuttuğumuz sınıfın kendi davranışını sergilemesini istiyoruz.
        super.dataSource.readData();
    }

    @Override
    public void writeData(Object data) {
        // Burada gelen veriye ait sıkıştırma işlemleri yapılacaktır.
        System.out.println("Data compressed.");

        // Sıkıştırma işlemleri yapıldıktan sonra yazma işlemleri gerçekleştirilecektir.
        System.out.println("Compressed data was written to " + super.dataSource.getFileName());
    }
}

File: decorator/python/comporession_decorator.py
from base_data_source_decorator import BaseDataSourceDecorator


class ComporessionDecorator(BaseDataSourceDecorator):
    """
    Yeni davranışların tanımlandığı sınıftır.
    UML diyagramındaki BaseDecorator (BaseDataSourceDecorator) sınıfından türer.
    UML diyagramındaki ConcreteDecorator sınıfına denk gelmektedir.
    """

    def get_file_name(self):
        return self.data_source.get_file_name()

    def write_data(self, data):
        # Burada gelen veriye ait sıkıştırma işlemleri yapılacaktır.
        print("Data compressed.")

        # Sıkıştırma işlemleri yapıldıktan sonra yazma işlemleri gerçekleştirilecektir.
        print(f"Compressed data was written to {self.data_source.get_file_name()}")

    def read_data(self):
        # Veriyi okurken referansını tuttuğumuz sınıfın kendi davranışını sergilemesini istiyoruz.
        self.data_source.read_data()
